Add pdfreport webservice

asaunier opened the pull request:
This webservice allows the users to get a full PDF report concerning a given feature (layername+id), using  MapFish Print 3 and some jasper reports.

WIP, still TODO:
* handling mapserver groups

sbrunner commented on c2cgeoportal/views/pdfreport.py:
```diff
@@ -0,0 +1,162 @@
+    def getreport(self):
+        id = int(self.request.matchdict['id'])
+        layername = self.request.matchdict['layername']
+        if id is None or layername is None:
```
Not possible ... :-)

asaunier replied:
OK I'll remove it.

sbrunner commented on c2cgeoportal/views/pdfreport.py:
```diff
@@ -0,0 +1,162 @@
+            'request': 'GetFeature',
+            'typeName': layername,
+            'featureid': featureid,
+            'outputFormat': 'geojson'
```
GeoJson ? 
I think you should add come documentation about this ...

asaunier replied:
Yes, this requires adding a "geojson" outputformat in the mapfile. Some doc is still needed for this part.

sbrunner replied:
I think that the MapFish Print support the GML, isn't it the case ?

asaunier replied:
> I think that the MapFish Print support the GML, isn't it the case ?

I don't know. If I replace
```
"geoJson": request_url
```
by
```
"gml": request_url
```
(with request_url being the WFS request with no output format - default is GML), the printing fails saying that:
```
Error while processing request:
org.mapfish.print.ExtraPropertyException: Extra properties were found in the request data at: spec.attributes.map.layers[0]: 
Extra Properties: 

	* gml

All allowed properties are: 

	* style
	* name
	* opacity
	* renderAsSvg
	* geoJson
```

tsauerwein replied:
mfp 3 supports GML, you'll have to set the layer type to `type: gml` and then give an `url`.

asaunier replied:
@tsauerwein Thanks for the tip, it indeed works perfectly :)

sbrunner replied:
Cool :-)

sbrunner commented on c2cgeoportal/views/pdfreport.py:
```diff
@@ -0,0 +1,162 @@
+            + "&".join(["%s=%s" % i for i in params.items()])
+
+        baselayer = self.config.get("layers", {}). \
+            get(layername, {}).get('baselayers')
```
layers.baselayers is a new settings ?

sbrunner replied:
You use layers namespace but it's normally reserved for layers view!
I think that should be 'pdfreport' or 'pdfreport.layers'

sbrunner replied:
in GeoMapFish 2.0 we prefer the therm 'backgroundlayers' instance of 'baselayers'.

asaunier replied:
Doc needed here as well. Here is an example of config in config.yaml.in:
```
pdfreport:
    mapserv_url: https://${host}/${vars:instanceid}/wsgi/mapserv_proxy
    print_url: http://localhost:8080/print-c2cgeoportal-v3-${vars:instanceid}
    default_baselayers: ['grp_ly_tilegenerierung_landeskarte']
    default_imageformat: 'image/png'
    srs_name: 'EPSG:21781'
    layers:
        ly_a020_belastete_standorte_point:
            baselayers: ['grp_ly_tilegenerierung_landeskarte','ly_a020_belastete_standorte_point']
```

sbrunner replied:
`mapserv_url` should be the same as `request.get_route('mapserv_proxy')`
`print_url` already exists can we reuse it ?

Otherwise that looks good :-)

asaunier replied:
> You use layers namespace but it's normally reserved for layers view!
I think that should be 'pdfreport' or 'pdfreport.layers'

That's already the case, see the config example above. In the service we have
```
self.config = self.request.registry.settings.get('pdfreport', {})
```

> print_url already exists can we reuse it ?

Not really because we have a project that use MF Print 3 for this pdfreport tool but still use MF Print 2 for the standard printing.

sbrunner replied:
OK, I miss that :-)

sbrunner commented on c2cgeoportal/views/pdfreport.py:
```diff
@@ -0,0 +1,162 @@
+        baselayer = self.config.get("layers", {}). \
+            get(layername, {}).get('baselayers')
+        if baselayer is None:
+            baselayer = self.config.get('default_baselayers')
```
also ?

sbrunner commented on c2cgeoportal/views/pdfreport.py:
```diff
@@ -0,0 +1,162 @@
+                    "projection": srs,
+                    "dpi": 254,
+                    "rotation": 0,
+                    "bbox": [0, 0, 1000000, 1000000],
```
This looks like a workaround... is it report i mapfish-print ?

asaunier replied:
I don't know, this is a part added by @gilliM

sbrunner commented on c2cgeoportal/views/pdfreport.py:
```diff
@@ -0,0 +1,162 @@
+                    "rotation": 0,
+                    "bbox": [0, 0, 1000000, 1000000],
+                    "zoomToFeatures": {
+                        "zoomType": "center",
```
shouldn't be bbox instance of center ?

sbrunner commented on c2cgeoportal/views/pdfreport.py:
```diff
@@ -0,0 +1,162 @@
+                                    "pointRadius": 10
+                                }]
+                            }
+                        },
```
I think that the style should be in the config

asaunier replied:
Actually in some projects the map would not be used. So I think that one should be able to pass the whole ``spec`` in the config.

sbrunner replied:
I agree, but it will be not easy to do ...

asaunier commented:
I have added a documentation and implemented the suggested improvements.
Please note that the full JSON spec may now be provided in the configuration if needed.

sbrunner commented on c2cgeoportal/views/pdfreport.py:
```diff
@@ -0,0 +1,169 @@
+                    }}
+                }}
+            }}
+            """
```
:( I relay don't like this template ..., why not using only in the case we specify it in the configuration ...
and with: `loads(dumps(self.config.get("spec")) % (...))` ?

I use the % to avoid issue with `{}`

asaunier replied:
>  why not using only in the case we specify it in the configuration 

In order to use the same consistent mechanism whereas a spec_template is provided or not: if it works for default template, it should work for a param template.

sbrunner replied:
Never mind do it as structured config in booth side and call every tine the command I purpose ...

sbrunner commented on c2cgeoportal/views/pdfreport.py:
```diff
@@ -0,0 +1,169 @@
+            'typeName': layername,
+            'featureid': featureid
+        }
+        srs = self.config.get('srs_name')
```
Shouldn't simply be `srs` ?

asaunier replied:
OK

asaunier commented:
* Fixed commits up
* Used JSON config as suggested by @sbrunner 
* Updated doc
* Added ``# pragma: no cover`` to disable testing because PDF output is difficult to test (as for the printproxy service)

sbrunner commented:
Excellent, that looks good to me :-)

asaunier commented:
OK, thanks for reviewing Steph!
